Moves calculations related to the monthly page to a new service

ashwin003 opened the pull request:
Closes #103

ashwin003 committed:
Moves calculations related to the monthly page to a new service

chrisK00 commented on subtrack.MAUI/Pages/Monthly.razor:
```diff
@@ -1,45 +1,32 @@
-    IEnumerable<SubscriptionsMonthResponse> _nextYearSubscriptionMonths = null!;
-    IEnumerable<SubscriptionsMonthResponse> _twoYearsForwardSubscriptionMonths = null!;
-
+    int currentYear;
```
Prefix with underscore for private fields

chrisK00 commented on subtrack.MAUI/Services/Abstractions/IMonthlyPageCalculator.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface IMonthlyPageCalculator
+    {
+        Task<IEnumerable<IGrouping<int, SubscriptionsMonthResponse>>> GetMonthlySubscriptionLists(DateTime fromIncludedDate, DateTime toIncludedDate);
```
This is a weird type to deal with. The submonthitem already contains the year and the items are not only grouped by year, they are grouped by SubMonth but we havent introduced a type that represents that. Would use a dict in that case.

I can look the other way if you just change the return type to dictionary<int, x> and add a comment on interface that the int stands for the year

chrisK00 commented on subtrack.MAUI/Services/MonthlyPageCalculator.cs:
```diff
@@ -0,0 +1,45 @@
+
+        public MonthlyPageCalculator(ISubscriptionsCalculator subscriptionsCalculator, ISubscriptionService subscriptionService)
+        {
+            this.subscriptionsCalculator = subscriptionsCalculator;
```
Underscore instead

chrisK00 commented on subtrack.MAUI/Services/MonthlyPageCalculator.cs:
```diff
@@ -0,0 +1,45 @@
+
+        public async Task<IEnumerable<IGrouping<int, SubscriptionsMonthResponse>>> GetMonthlySubscriptionLists(DateTime fromIncludedMonthDate, DateTime finalIncludedMonthDate)
+        {
+            var subscriptions = await subscriptionService.GetAllAsync();
```
This an unnecessary dependency

ashwin003 replied:
please elaborate.

chrisK00 replied:
Instead of having a dependency on isubserv we can let the caller pass in the subs. That way we also dont force a reload of subs in case caller already has the subs. 

Input/output calls should always be centralized and the caller should be responsible for doing them

ashwin003 replied:
We have, and will have only one caller for this service.

chrisK00 replied:
Unnecessary risk to take. Cant be sure of that and it even makes testing worse as we have to mock the subscriptionservice

There is no pro of adding this line of code, only cons

Less dependencies=less chance of code breaking and less need of maintenance. E.g what if the caller wants to be able to filter? Are we now supposed to copy the params of getsubs service method?

ashwin003 replied:
We'll refactor if it comes to that. No point in anticipating future needs that may or may not materialize.

chrisK00 replied:
> We'll refactor if it comes to that. No point in anticipating future needs that may or may not materialize.

True but its a highly likely case. Limiting your caller, forcing them to refactor when you could have just left a simple parameter in the params doesnt make sense, why would you consider adding a dependency is good design? I dont see any benefits of doing so unless it simplified something for the caller or is required to due business logic, which it does not. It just adds magic . When designing a system you shouldnt aim for more coupling, you should aim for less coupling its software dev 1:1. I can recommend the book Clean Architecture if you are more interested on the matter

ashwin003 replied:
I disagree but I'll make the change if you feel this strongly about it.

chrisK00 replied:
You havent told me why you think its a good idea to add the dependency

ashwin003 replied:
At the moment, we have only one caller (the one we refactored for), and as far as the caller is concerned, it needs a list of subscriptions grouped by a certain criteria. This makes the interface a lot simpler. 

Ideally I would have also made it so that all the caller needs to pass the number of years of future data which is needed (2 in our case) and the service would handle the rest. 

This of course, assumes that the service would be used exclusively by this caller so it does call for a refactor if/when that changes.

ashwin003 replied:
I've run into a major issue, and will need you to make this change

 ![IMG_20230822_120258969.jpg](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/22453573-476e-435b-be10-2163745982dc)

chrisK00 replied:
I can understand you see it that way, but the interface doesnt really get simpler - it gets more complicated as the app grows because the subs are magically fetched by this method. If we centralized the fetching then they can be reused for example for other calculations (which is not entirely impossible for this view, maybe next to year we decide to show the total cost for only that year) and we could do so without having to modify the source method - which allows for Open/Closed principle in SOLID.

Restricting input I can understand but at the same time it does not add any value of doing so. But not doing so does add value because we allow them to get from a specific month to a specific month and thats another not so impossible case we might have. Instead of blindly saying you can see 2 years in the future, maybe we say you can see untill the last sub is due f.e if you have yearly every 3 years or whatever.

Adding that line of code doesnt simplify anything. Its only one line of code to fetch subs. If it was something complicated then ok I would agree with you but there is no business logic involved when fetching a sub. 

It does make things harder, limits the caller to reuse subs, limits them to filter, forces them to provide ISubService, forces our tests to mock it and the caller has no idea where we fetch the subs from without having to pay us a visit.

Also I have not ruled out that this will be the only caller in the future since this can be used for some interesting things like statistics without having to add more unnecessary methods.

chrisK00 replied:
> I've run into a major issue, and will need you to make this change
> 
>  ![IMG_20230822_120258969.jpg](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/22453573-476e-435b-be10-2163745982dc)
> 
> 

Rip lol. Ill do it after work

chrisK00 replied:
Postponing this to the weekend, in the process of buying an apartment

chrisK00 commented on subtrack.MAUI/Pages/Monthly.razor:
```diff
@@ -1,45 +1,32 @@
+    {
+        <h6 class="text-center mt-3 mb-2">@(subs.Key)</h6>
+    }
+    @foreach(var subMonth in subs)
```
Good cleanup #scoutRule

ashwin003 committed:
Refactorings:
private field naming convention
Dictionary instead of IGrouing returned from the service

chrisK00 committed:
merge dev

chrisK00 committed:
undo commit
